conan/tools/gnu/autotools.py

```diff
@@ -0,0 +1,239 @@
+                    args.append("--datarootdir=${prefix}/%s" % DEFAULT_SHARE)
+
+        with environment_append(pkg_env):
+            with environment_append(vars or self.vars):
+                command = '%s/configure %s %s' % (configure_dir, args_to_string(args),
+                                                  " ".join(triplet_args))
+                self._conanfile.output.info("Calling:\n > %s" % command)
+                self._conanfile.run(command, win_bash=self._win_bash, subsystem=self.subsystem)"""
+
+        cmd = "bash -c 'source autotoolsdeps.sh && source autotools.sh && %s/configure'" % configure_dir
+        self._conanfile.output.info("Calling:\n > %s" % cmd)
+        self._conanfile.run(cmd, win_bash=self._win_bash, subsystem=self.subsystem)
+
+    def _configure_help_output(self, configure_path):
+        from six import StringIO  # Python 2 and 3 compatible
+        mybuf = StringIO()
+        try:
+            self._conanfile.run("%s/configure --help" % configure_path, win_bash=self._win_bash,
+                                output=mybuf)
+        except ConanException as e:
+            self._conanfile.output.warn("Error running `configure --help`: %s" % e)
+            return ""
+        return mybuf.getvalue()
+
+    def _adjust_path(self, path):
+        if self._win_bash:
+            path = unix_path(path, path_flavor=self.subsystem)
+        return '"%s"' % path if " " in path else path
+
+    @staticmethod
+    def _valid_configure_flag(varname, args, available_flags):
+        return not AutoToolsBuildEnvironment._is_flag_in_args(varname, args) and \
+               varname in available_flags
+
+    @staticmethod
+    def _is_flag_in_args(varname, args):
+        flag = "--%s=" % varname
+        return any([flag in arg for arg in args])
+
+    def make(self, target=None):
+        """if not self._build_type:
+            raise ConanException("build_type setting should be defined.")
+        with environment_append(vars or self.vars):
+            str_args = args_to_string(args)
+            cpu_count_option = (("-j%s" % cpu_count(output=self._conanfile.output))
+                                if ("-j" not in str_args and "nmake" not in make_program.lower())
+                                else None)
+            self._conanfile.run("%s" % join_arguments([make_program, target, str_args,
+                                                       cpu_count_option]),
+                                win_bash=self._win_bash, subsystem=self.subsystem)"""
+
+        make_program = self._conanfile.conf["tools.gnu"].make_program
+        if make_program is None:
+            make_program = "mingw32-make" if platform.system() == "Windows" else "make"
+        if platform.system() == "Windows":
+            cmd = "autotoolsdeps.bat && autotools.bat && {}".format(make_program)
+        else:
+            cmd = "bash -c 'source autotoolsdeps.sh "\
```

Comment:
I think we shouldn't hard-code bash here

Reply:
Unless it is the only way to make it work reproducibly. The autotoolsdeps.sh, etc are using bash. If we are supporting other shells, we need tests, I don't want to add things into the new codebase that cannot be tested.

Reply:
the make command is not hard-coded and got from the config.
the same, more or less, should be done for the bash I believe.
also, maybe it's enough to just use plain `sh` or `$SHELL`?
I don't think we need advanced bash-specific features, and it would be nice to be portable and require only posix sh.
bash is not always available and not a default shell everywhere.

Reply:
The ``make`` command is hardcoded, most of the times it doesn't come from config: https://github.com/conan-io/conan/blob/2ccc79c48ce42972a19c7d3587e4ab8b585f220a/conans/client/build/autotools_environment.py#L237

I guess you are aware that many recipes in ConanCenter are doing ``make_program=which("make") or which("mingw32-make")`` to be able to work on Windows. No config here, but recipes hardcoding it.

Same as ``bash``, hardcoded here: https://github.com/conan-io/conan/blob/2ccc79c48ce42972a19c7d3587e4ab8b585f220a/conans/client/tools/oss.py#L389.

The fact that they have configs, doesn't mean that they are not hardcoded in the codebase. Same I am doing here, we can discuss some defaults as we gain knowledge, but I am not introducing new hardcoded things.

The new ``Environment`` files at the moment only work on ``bash``. If we can generalize them to work on different shells, or to provide different generated files, for different shells, good for me.

Reply:
I know there are recipes doing that, as I wrote some of them.
that might be a sign we're not managing things the best way.
in some cases, `mingw32-make` is valid defaults, in some cases valid default could be `nmake` or just plain `make`.
that might be the reason we need to provide helper to deduce sane default based on settings, e.g. it may return MinGW make, nmake, GNU make or BSD make depending on system and compiler.
but that we know already the logic doesn't believe  to the AutoToolsBuildEnvironment or recipes.

Reply:
Yes, agree, a helper would be the best. This helper can hardcode some defaults for some given configurations (settings, etc), and always have a ``[conf]`` to be able to override it. 

But lets wait a bit and have a minimum set of build helpers, just the most common "AutotoolsXXX" ones, that work well and make recipes simpler for a majority of cases, before extracting new helpers like ``get_make_program(conanfile)``, when we gain more insights about these defaults, our new helpers will be closer to something stable.